src/slot_provider/clock.rs

```diff
@@ -1,88 +1,180 @@
 //! Clock based SlotProvider
-//! When a fn does not finish within a slot (12 seconds), the next call is made concurrently, up to a buffer limit. When the buffer limit is reached, the passed fn will not be called for a new slot, but wait in a FIFO queue until a previous slot has finished.
```

Comment:
My Summary of the high level changes in the PR. (posting this here because top level PR comments are clunky and don't allow threads).

Old State:
One `IntervalStream` executes every `SLOT_PERIOD` seconds determines the slot number "on the fly" and processes the slot.

New State:
We have two streams:
1. `IntervalStream` that gets triggered every `SLOT_PERIOD` seconds, determines current slot and pushes it onto the stack and "pings" the channel
2. `ReceiverStream` gets triggered by the ping from the first stream, pops the last slot off the queue and processes it.

Is this understanding correct ? I

Reply:
Assuming above understanding to be correct it seems that the functionality would change in the case of "overload" (not being able to handle enough slots in time).

Whereas the original state would skip slots but make sure every slot is handled at the correct time, the new state seems to potentially fall behind, i.e. introducing a delay between the actual slot time (when it was pushed onto the queue) and when it is handled.

Whereas I remember verifying the original behaviour, the statement about the current behaviour is just based on my reading of the code.

Can you confirm this ? Have you tested it ?
If above is true I think it might be a problem because prices might be outdated / wrong when we submit them.

Reply:
This is related to the following comments:

https://github.com/ultrasoundmoney/oracle-client/pull/16#discussion_r1253912210
https://github.com/ultrasoundmoney/oracle-client/pull/16#discussion_r1253913696

Reply:
With some testing (setting slot time to 6s and max concurrency to 1)  I was able to produce a situation where a slots start processing 8 seconds after the slot period has started:

<img width="1438" alt="image" src="https://github.com/ultrasoundmoney/oracle-client/assets/15629702/4b023591-44c7-433d-bd38-b56798ecff15">

Reply:
Yea that's how I built it but I agree that's not ideal. Let's figure out here what we want. I also feel the following is better actually:
>One IntervalStream executes every SLOT_PERIOD seconds determines the slot number "on the fly" and processes the slot.

So agreeing that slots can actually be figured out on the fly, we still have to figure out how the AttestationScheduler schedules work. I'd say actually, v1, let's forget about leap seconds. We:
* start
* wait for the next slot to start
* start a 12s interval and schedule on top of that

Would look much more like what you have already.

On top of that, for the scheduling, I propose the following:
* we start processing a new slot at the start of the interval
* after 24s, we consider the slot too stale for any message broadcasting, and give up our efforts

This would mean concurrency is automatically limited to 2, although we could at some point add a hard concurrency limit.

How does that sound?

Reply:
I went ahead and implemented the above while the ideas were fresh in my mind. You can see the new code in the PR. Looks quite different!

Reply:
Implementation in this commit looks good to me: https://github.com/ultrasoundmoney/oracle-client/pull/16/commits/491ab47a11abae5119a198494b2e9570a1b5bbbf

Great stuff.👏

Reply:
I like how this version combines a reduction in the type noise / syntactical complexity while keeping the execution logic simple. (no queue, single stream).